presto-native-execution/presto_cpp/main/operators/StorageBasedLocalShuffle.cpp

```diff
@@ -0,0 +1,128 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "presto_cpp/main/operators/StorageBasedLocalShuffle.h"
+#include <filesystem>
+
+using namespace facebook::velox::exec;
+using namespace facebook::velox;
+
+namespace {
+inline std::string
+createShuffleFilename(std::string rootPath, int32_t partition, int fileIndex) {
+  return rootPath + "/" + std::to_string(partition) + "_" +
+      std::to_string(fileIndex) + ".bin";
+}
+}; // namespace
+
+namespace facebook::presto::operators {
+
+std::unique_ptr<velox::WriteFile> StorageBasedLocalShuffle::getNextOutputFile(
+    int32_t partition) {
+  auto fileCount = getPartitionFilesCount(partition);
+  std::string filename = createShuffleFilename(rootPath_, partition, fileCount);
+  return fileSystem_->openFileForWrite(filename);
+}
+
+bool StorageBasedLocalShuffle::fileExists(const std::string& filename) const {
+  if (fileSystem_->name() == "Local FS") {
+    return std::filesystem::exists(filename);
+  }
+  VELOX_FAIL("Only supports local filesystem.");
+}
+
+int StorageBasedLocalShuffle::getPartitionFilesCount(int32_t partition) const {
+  auto fileCount = 0;
+  bool fileExist = false;
+  do {
+    std::string filename =
+        createShuffleFilename(rootPath_, partition, fileCount);
+    fileExist = fileExists(filename);
+    if (!fileExist) {
+      break;
+    }
+    fileCount++;
+  } while (fileExist);
+  return fileCount;
+}
+
+void StorageBasedLocalShuffle::storePartitionBlock(int32_t partition) {
+  auto& buffer = inProgressPartitions_[partition];
+  buffer->setSize(inProgressSizes_[partition]);
+  auto file = std::move(getNextOutputFile(partition));
+  file->append(std::string_view(buffer->asMutable<char>(), buffer->size()));
+  file->close();
+  inProgressPartitions_[partition].reset();
+}
+
+void StorageBasedLocalShuffle::collect(
+    int32_t partition,
+    std::string_view data) {
+  auto& buffer = inProgressPartitions_[partition];
+
+  // Check if there is enough space in the buffer.
+  if (buffer &&
+      inProgressSizes_[partition] + data.size() + sizeof(size_t) >=
+          maxBytesPerPartition_) {
+    storePartitionBlock(partition);
+  }
+
+  // Allocate buffer if needed.
+  if (!buffer) {
+    buffer = AlignedBuffer::allocate<char>(maxBytesPerPartition_, pool_);
+    inProgressSizes_[partition] = 0;
+  }
+
+  // Copy data.
+  auto rawBuffer = buffer->asMutable<char>();
+  auto offset = inProgressSizes_[partition];
+
+  *(size_t*)(rawBuffer + offset) = data.size();
+
+  offset += sizeof(size_t);
+  memcpy(rawBuffer + offset, data.data(), data.size());
+
+  inProgressSizes_[partition] += sizeof(size_t) + data.size();
```

Comment:
We may not need to maintain inProgressSizes_ if we set the size on inProgressSizes_[partition]

```
inProgressSizes_[partition]->setSize(offset + data.size());
```

Reply:
Yeah that's a great idea. I was thinking about the same. Less state is much better

Reply:
Just to be sure you mean setting the size on inProgress §§§inProgressPartitions_[partition]§§§

Reply:
Yes, inProgressPartitions_